**<font color = black size=6>实验十:EM算法</font>**

In [1]:
from collections import Counter
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import math
import random

**<font color = blue size=4>第一部分:实验任务</font>**

<span style="color:purple">=======  
现在给定了一个班级中所有同学的身高数据文件(height.csv)，但不知道各个学生的性别。假设男生身高服从一高斯分布$N_1(\mu_1,\sigma_1^2)$，女生身高服从另一高斯分布$N_2(\mu_2,\sigma_2^2)$，我们可以写出一个混合高斯模型:$x\sim\alpha_1 N_1(\mu_1,\sigma_1^2)+\alpha_2 N_2(\mu_2,\sigma_2^2)$。请使用EM算法完成对该混合高斯分布的估计(即求出对参数$\alpha_1,\alpha_2,\mu_1,\sigma_1,\mu_2,\sigma_2$的估计值)。我们简化地记$\theta_1=(\alpha_1,\mu_1,\sigma_1)$, $\theta_2=(\alpha_2,\mu_2,\sigma_2)$</span>

<span style="color:purple">该数据集(height.csv)特征信息只包括了1个特征，即学生的身高。我们沿用理论课PPT的设置，将隐变量$z_i$按照男生$z_i=1$、女生$z_i=2$的形式进行描述</span>

**<font color = black size=4>E步（Expectation Step）</font>**

<span style="color:purple">1) 将数据集'height.csv'载入并转换为你需要的格式</span>

In [29]:
D = pd.read_csv('data.csv')
for i in D["height"]:
    print(i)

170.2693075091544
179.88604529847407
173.6813290631826
179.8124729998782
174.8148166898798
177.99077689337548
175.62870511737736
177.7396063405258
175.270333360301
178.27679368674677
171.38637233020782
176.85962173020334
174.218494170901
172.9724116675917
166.02548975662882
173.53569624455943
168.95490432063218
175.45962347567794
175.40067941201767
171.8303016514757
174.40836383338498
170.13338685030476
175.54804192314697
175.15357527655496
168.87456892869648
170.77202170905915
171.3641694984213
175.96098877597234
180.01894946567523
184.14900337335655
180.99531665417305
173.42994199125474
178.92066029051165
170.71613176113084
179.61008266474482
172.85299836003128
168.8945324622705
168.95694842009704
169.23521314391832
187.1540434716583
172.6525022567047
173.23182821393604
176.89224164741552
171.3636598104152
175.48729684278854
176.8612695629201
180.004783295648
167.98026615279622
177.99403911477611
174.2888518263463
181.755236500169
171.0365126034487
172.5092769357089
174.7408275328298

<span style="color:purple">2)初始化  
初始化$t=0$时的参数($\alpha_1(0)$, $\alpha_2(0)$, $\mu_1(0)$, $\mu_2(0)$, $\sigma_1(0)$, $\sigma_2(0)$). </span>

In [30]:
alpha1 = 0.5
alpha2 = 0.5
mu1 = 170
mu2 = 170
sigma1 = 4
sigma2 = 4
parameter = [alpha1, alpha2, mu1, mu2, sigma1, sigma2]

<span style="color:purple">3)编写函数P(x,parameter,z)  
给定参数$(\alpha_1(t),\alpha_2(t),\mu_1(t),\mu_2(t),\sigma_1(t),\sigma_2(t))$以及数据集D,计算每个样本$x_i$的$P(x_i,z_i|\theta)$.</span>

<span style="color:purple">.  
当$z_i=1$时,$$P(x_i,z_i|\theta)=\alpha_1(t)f_1(x_i|\theta_1(t))$$
    当$z_i=2$时,$$P(x_i,z_i|\theta)=\alpha_2(t)f_2(x_i|\theta_2(t))$$</span>

<span style="color:purple"> .   
其中$f_1(x_i|\theta_1(t))$为样本$x_i$在模型$N_1$中的概率密度,公式如下:
    $$f(x_i|\theta_1(t))=\frac{1}{{\sqrt{2\pi}\sigma_1}} e^{-\frac{(x_i-\mu_1)^2}{2\sigma_1^2}}$$</span>

In [31]:
#parameter = [alpha1 0, alpha2 1, mu1 2, mu2 3, sigma1 4, sigma2 5]
def P(x, parameter, z):
    if z==1:
        return parameter[0] * np.exp(-(x - parameter[2])**2 / (2*parameter[4]**2)) / (np.sqrt(2*np.pi) * parameter[4])  
    elif z==2:
        return parameter[1] * np.exp(-(x - parameter[3])**2 / (2*parameter[5]**2)) / (np.sqrt(2*np.pi) * parameter[5])
'''
def main():
    # 定义测试参数
    test_parameter = [0.5, 0.7, 1.0, 2.0, 0.5, 1.0]

    # 测试条件 z=1
    x_value = 1.5
    result_z1 = P(x_value, test_parameter, z=1)
    expected_result_z1 = test_parameter[0] * np.exp(-(x_value - test_parameter[2])**2 / (2 * test_parameter[4]**2)) / (np.sqrt(2 * np.pi) * test_parameter[4])
    assert np.isclose(result_z1, expected_result_z1, atol=1e-8)

    # 测试条件 z=2
    x_value = 3.0
    result_z2 = P(x_value, test_parameter, z=2)
    expected_result_z2 = test_parameter[1] * np.exp(-(x_value - test_parameter[3])**2 / (2 * test_parameter[5]**2)) / (np.sqrt(2 * np.pi) * test_parameter[5])
    assert np.isclose(result_z2, expected_result_z2, atol=1e-8)

    print("All tests passed!")

if __name__ == "__main__":
    main()
'''

P1 = P(D,parameter,1)
print(P1)

        height
0     0.049755
1     0.002352
2     0.032651
3     0.002461
4     0.024165
...        ...
1995  0.046489
1996  0.049458
1997  0.013652
1998  0.013286
1999  0.044985

[2000 rows x 1 columns]


<span style="color:purple">4)编写函数Y(x,parameter,z)  
给定参数$(\alpha_1(t),\alpha_2(t),\mu_1(t),\mu_2(t),\sigma_1(t),\sigma_2(t))$以及数据集D,计算每个样本$x_i$的$y_{1,i}=P((z_i=1)|x_i,\theta)$和$y_{2,i}=P((z_i=2)|x_i,\theta)$.  
公式如下:  
</span>

$$P((z_i=1)|x_i,\theta) = \frac{\alpha_1(t)f_1(x_i|\theta_1(t))}{\alpha_1(t)f_1(x_i|\theta_1(t))+\alpha_2(t)f_2(x_i|\theta_2(t))}$$  
$$P((z_i=2)|x_i,\theta) = \frac{\alpha_2(t)f_2(x_i|\theta_2(t))}{\alpha_1(t)f_1(x_i|\theta_1(t))+\alpha_2(t)f_2(x_i|\theta_2(t))}$$

In [6]:
#parameter=[alpha1 0,alpha2 1,mu1 2,mu2 3,sigma1 4,sigma2 5]
def Y(x,parameter,z):
    if z==1:
        return P(x, parameter, 1) / (P(x, parameter, 1) + P(x, parameter, 2))
    elif z==2:
        return P(x, parameter, 2) / (P(x, parameter, 1) + P(x, parameter, 2))
Y1 = Y(D,parameter,1)
print(Y1)

      height
0        0.5
1        0.5
2        0.5
3        0.5
4        0.5
...      ...
1995     0.5
1996     0.5
1997     0.5
1998     0.5
1999     0.5

[2000 rows x 1 columns]


<span style="color:purple">5)编写函数Q(x,parameter)  
 计算对数似然函数在该分布和基于$\theta(t)$下的期望值$Q(\theta)$.单个样本的期望值公式如下:$$E_{z_i}logP(x_i,z_i|\theta)=y_{1,i}log(\alpha_1(t)f_1(x_i|\theta_1(t)))+y_{2,i}log(\alpha_2(t)f_2(x_i|\theta_2(t)))$$</span>

In [60]:
#parameter=[alpha1,alpha2,mu1,mu2,sigma1,sigma2]
def Q(x, parameter):
    return Y(x, parameter, 1) * np.log(P(x, parameter, 1)) + Y(x, parameter, 2) * np.log(P(x, parameter, 2))

**<font color = black size=4>M步 (Maximization Step)</font>**

<span style="color:purple">6)编写函数alpha_expection(D,parameter)  
 给定参数$(\alpha_1(t),\alpha_2(t),\mu_1(t),\mu_2(t),\sigma_1(t),\sigma_2(t))$以及数据集D，计算第$(t+1)$轮的$(\alpha_1(t+1),\alpha_2(t+1))$的更新值.
</span>

$$\alpha_1(t+1)=\frac{\sum_{i=1}^m{y_{1,i}}}{m}$$  
$$\alpha_2(t+1)=\frac{\sum_{i=1}^m{y_{2,i}}}{m}$$

In [32]:
#parameter=[alpha1,alpha2,mu1,mu2,sigma1,sigma2]
def alpha_expection(D, parameter):
    n_samples, _ = D.shape
    alpha1_new = 0
    alpha2_new = 0
    for height in D["height"]:
        alpha1_new += Y(height, parameter, 1)
        alpha2_new += Y(height, parameter, 2)
    alpha1_new /= n_samples
    alpha2_new /= n_samples
    parameter[0] = alpha1_new
    parameter[1] = alpha2_new
    print(parameter)
    return parameter
p = alpha_expection(D,parameter)

[0.5, 0.5, 170, 170, 4, 4]


<span style="color:purple">7)编写函数mu_expection(D,parameter)  
给定参数$(\alpha_1(t),\alpha_2(t),\mu_1(t),\mu_2(t),\sigma_1(t),\sigma_2(t))$以及数据集D，计算第$(t+1)$轮的$(\mu_1(t+1),\mu_2(t+1))$的更新值.
</span>

$$\mu_1(t+1)=\frac{\sum_{i=1}^m{y_{1,i}x_i}}{\sum_{i=1}^m{y_{1,i}}}$$
$$\mu_2(t+1)=\frac{\sum_{i=1}^m{y_{2,i}x_i}}{\sum_{i=1}^m{y_{2,i}}}$$

In [33]:
#parameter=[alpha1,alpha2,mu1,mu2,sigma1,sigma2]
def mu_expection(D, parameter):
    mu1_new = 0
    mu2_new = 0
    denominator1 = 0
    denominator2 = 0
    for height in D["height"]:
        mu1_new += Y(height, parameter, 1) * height
        mu2_new += Y(height, parameter, 2) * height
        denominator1 += Y(height, parameter, 1)
        denominator2 += Y(height, parameter, 2)
    mu1_new /= denominator1
    mu2_new /= denominator2
    parameter[2] = mu1_new
    parameter[3] = mu2_new
    print(parameter)
    return parameter
p = mu_expection(D,parameter)

[0.5, 0.5, 171.54328965959186, 171.54328965959186, 4, 4]


<span style="color:purple">8)编写函数sigma_expection(D,parameter,mu_next_1,mu_next_2)  
给定参数$(\alpha_1(t),\alpha_2(t),\mu_1(t),\mu_2(t),\sigma_1(t),\sigma_2(t))$以及数据集D，计算第$(t+1)$轮的$(\sigma_1(t+1),\sigma_2(t+1))$的更新值.
</span>

$$\sigma_1(t+1)=\sqrt{\frac{\sum_{i=1}^m{y_{1,i}(x_i-\mu_1(t+1))^2}}{\sum_{i=1}^m{y_{1,i}}}}$$
$$\sigma_2(t+1)=\sqrt{\frac{\sum_{i=1}^m{y_{2,i}(x_i-\mu_2(t+1))^2}}{\sum_{i=1}^m{y_{2,i}}}}$$

In [34]:
#parameter=[alpha1,alpha2,mu1,mu2,sigma1,sigma2]
def sigma_expection(D, parameter):
    sigma1_new = 0
    sigma2_new = 0
    denominator1 = 0
    denominator2 = 0
    for height in D["height"]:
        sigma1_new += Y(height, parameter, 1) * (height - parameter[2])**2
        sigma2_new += Y(height, parameter, 2) * (height - parameter[3])**2
        denominator1 += Y(height, parameter, 1)
        denominator2 += Y(height, parameter, 2)
    sigma1_new /= denominator1
    sigma2_new /= denominator2
    sigma1_new = np.sqrt(sigma1_new)
    sigma2_new = np.sqrt(sigma2_new)
    parameter[4] = sigma1_new
    parameter[5] = sigma2_new
    print(parameter)
    return parameter
p = sigma_expection(D,parameter)

[0.5, 0.5, 171.54328965959186, 171.54328965959186, 6.8566064401359625, 6.8566064401359625]


**<font color = black size=4>E步和M步的迭代过程</font>**

<span style="color:purple">9) 利用前面编写的函数完成EM算法的迭代过程，直至达到收敛要求。请至少完成【3次】不同的初始值下的迭代过程，并比较选出最好的。  
    收敛要求给出如下几种参考:  
    1.迭代轮数达到指定轮数;  
    2.每轮参数更新的差值小于阈值.</span>

<img src='./EM Algorithm Pseudocode.png'>

.  
我们给出这个数据集的正确相关信息作为参考:$\theta_1:(\alpha_1=0.625,\mu_1=175,\sigma_1=4)$,$\theta_2:(\alpha_2=0.375,\mu_2=165,\sigma_2=6)$

In [37]:
# 1
alpha1 = 0.6
alpha2 = 0.4
mu1 = 172
mu2 = 168
sigma1 = 5
sigma2 = 5
parameter = [alpha1, alpha2, mu1, mu2, sigma1, sigma2]
for _ in range(100):
    parameter = alpha_expection(D, parameter)
    parameter = mu_expection(D, parameter)
    parameter = sigma_expection(D, parameter)
print(parameter)

'''
# 2
alpha1 = 0.5
alpha2 = 0.5
mu1 = 170
mu2 = 170
sigma1 = 5
sigma2 = 5
parameter = [alpha1, alpha2, mu1, mu2, sigma1, sigma2]
for _ in range(100):
    parameter = alpha_expection(D, parameter)
    parameter = mu_expection(D, parameter)
    parameter = sigma_expection(D, parameter)
print(parameter)

# 3
alpha1 = 0.9
alpha2 = 0.1
mu1 = 180
mu2 = 160
sigma1 = 9
sigma2 = 1
parameter = [alpha1, alpha2, mu1, mu2, sigma1, sigma2]
for _ in range(100):
    parameter = alpha_expection(D, parameter)
    parameter = mu_expection(D, parameter)
    parameter = sigma_expection(D, parameter)
print(parameter)
'''


[0.6345065056172756, 0.36549349438272444, 172, 168, 5, 5]
[0.6345065056172756, 0.36549349438272444, 173.7698992941187, 167.2048878726735, 5, 5]
[0.6345065056172756, 0.36549349438272444, 173.7698992941187, 167.2048878726735, 4.969674698139274, 6.463438006602113]
[0.6349146961799075, 0.3650853038200926, 173.7698992941187, 167.2048878726735, 4.969674698139274, 6.463438006602113]
[0.6349146961799075, 0.3650853038200926, 174.36556167122205, 166.62924694778087, 4.969674698139274, 6.463438006602113]
[0.6349146961799075, 0.3650853038200926, 174.36556167122205, 166.62924694778087, 4.718463095665672, 6.7350910206659576]
[0.6304254678814105, 0.3695745321185917, 174.36556167122205, 166.62924694778087, 4.718463095665672, 6.7350910206659576]
[0.6304254678814105, 0.3695745321185917, 174.6699712812433, 166.27239627938664, 4.718463095665672, 6.7350910206659576]
[0.6304254678814105, 0.3695745321185917, 174.6699712812433, 166.27239627938664, 4.521581209094646, 6.644513199112869]
[0.6252633250719778, 0.37

'\n# 2\nalpha1 = 0.5\nalpha2 = 0.5\nmu1 = 170\nmu2 = 170\nsigma1 = 5\nsigma2 = 5\nparameter = [alpha1, alpha2, mu1, mu2, sigma1, sigma2]\nfor _ in range(100):\n    parameter = alpha_expection(D, parameter)\n    parameter = mu_expection(D, parameter)\n    parameter = sigma_expection(D, parameter)\nprint(parameter)\n\n# 3\nalpha1 = 0.9\nalpha2 = 0.1\nmu1 = 180\nmu2 = 160\nsigma1 = 9\nsigma2 = 1\nparameter = [alpha1, alpha2, mu1, mu2, sigma1, sigma2]\nfor _ in range(100):\n    parameter = alpha_expection(D, parameter)\n    parameter = mu_expection(D, parameter)\n    parameter = sigma_expection(D, parameter)\nprint(parameter)\n'

**<font color = blue size=4>第二部分:作业提交</font>**

一、实验课下课前提交完成代码，如果下课前未完成，请将已经完成的部分进行提交，未完成的部分于之后的实验报告中进行补充  
要求:  
1)文件格式为：学号-姓名.ipynb  
2)【不要】提交文件夹、压缩包、数据集等无关文件，只需提交单个ipynb文件即可，如果交错请到讲台前联系助教，删掉之前的错误版本后再进行提交

二、实验报告截止日期： 【11月24日 14:20】
要求：  
1)文件格式为：学号-姓名.pdf  
2)【不要】提交文件夹、压缩包、代码文件、数据集等任何与实验报告无关的文件，只需要提交单个pdf文件即可  
3)文件命名时不需要额外添加“实验几”等额外信息，按照格式提交  
4)每周的实验报告提交地址会变化，且有时间限制，提交时间为下周的实验课开始时，请注意及时提交。

实验十(EM算法)的实验报告上交地址:https://send2me.cn/9UjusmMn/S_Cz3o_FpKQEsA  

三、课堂课件获取地址:https://www.jianguoyun.com/p/DZKTh-IQp5WhChiIn6gFIAA  
实验内容获取地址:https://www.jianguoyun.com/p/DbCHB9wQp5WhChiKn6gFIAA